### PR TITLE
Umls relations

### PR DESCRIPTION
**Pull request by khituras**
This commit introduces the facilities to get taxonomical related concepts and their synonyms for addition to the query from the UMLS.

@michelole I noticed that currently, there is only a `hypernym` facility on the topic. Shouldn't there by `hyponyms`, too? Would be trivial to implement given the contents of this PR.

Please check if this looks OK.

### CODE REVIEW

**Comment by michelole** [comment]
Yes, we definitely should have hyponyms; Lexigram API "descendants" endpoint is currently broken and that might have also been the case in 2017.

**Comment by michelole** [comment]
> I am currently not sure if the results I get are actually hypernyms.

You could try to reproduce the cases in the `LexigramTest`.

File: scripts/createUmlsRelations.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		cui2    = line[4]
+		idType2 = line[6]
+		rel     = line[3]
+		suppressed = line[14]
+		# Filter for CUI id references because those are the concept relations.
+		if rel in eligibleRelations and suppressed == "N" and idType1 == "CUI" and idType2 == "CUI":
+			yield "\t".join([cui1, rel, cui2])
+
+def main():
+	mrrel = argv[1]
```
Comment (michelole) [inline scripts/createUmlsRelations.py]: What should be argv[1] here? The `MRREL.RRF` file?

Comment (khituras) [reply scripts/createUmlsRelations.py]: Yes. Yeah, sorry, I did forget to add a call example to the file.

File: src/main/java/de/julielab/ir/umls/UmlsRelationsProvider.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    private static final Logger log = LogManager.getLogger();
+    private static UmlsRelationsProvider instance;
+    private final String umlsSynsetFile;
+    private final String separator;
+    private static boolean useCache;
+    private CacheAccess<String, Set<String>> parentsCache;
+    private CacheAccess<String, Set<String>> childrenCache;
+    private static String defaultRelationsFile = "resources/umlsRelations.txt.gz";
+
+    public UmlsRelationsProvider(String umlsSynsetFile, boolean useCache) throws IOException {
```
Comment (michelole) [inline src/main/java/de/julielab/ir/umls/UmlsRelationsProvider.java]: Couldn't this be private as well, to force the singleton?

File: src/test/java/at/medunigraz/imi/bst/trec/query/DiseaseUmlsSynonymQueryDecoratorTest.java [hunk 3]
```diff
@@ -4,22 +4,24 @@
...
+        UmlsSynsetProvider.setUseCache(false);
         DummyElasticSearchQuery dummyQuery = new DummyElasticSearchQuery();
         final DiseaseUmlsSynonymQueryDecorator decorator = new DiseaseUmlsSynonymQueryDecorator(dummyQuery);
         final TopicSet topicSet = new TopicSet(new File("src/main/resources/topics/topics2018.xml"), Challenge.TREC_PM, Task.PUBMED, 2017);
         for (Topic topic : topicSet.getTopics()) {
             decorator.expandTopic(topic);
-            System.out.println(topic.getDisease());
-            System.out.println("   " + topic.diseaseSynonyms);
+            if (topic.getDisease().equals("melanoma"))
+                assertThat(topic.getDiseaseSynonyms()).contains("nevocarcinoma", "cutaneous melanoma", "melanoma (disorder)");
```
Comment (michelole) [inline src/test/java/at/medunigraz/imi/bst/trec/query/DiseaseUmlsSynonymQueryDecoratorTest.java]: Lexigram provides here "melanoblastoma", "melanosarcoma", and "melanocarcinoma".

Note that we should filter out "(disorder)" somewhere. There was e.g. some filtering in `Lexigram`.

File: src/test/java/at/medunigraz/imi/bst/trec/query/DiseaseUmlsHypernymQueryDecoratorTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        // We must also set the synset provider for testing because it is internally used
+        UmlsSynsetProvider.setSynsetSourceFile("src/test/resources/umls/melanoma.synset");
+        UmlsSynsetProvider.setUseCache(false);
+        DummyElasticSearchQuery dummyQuery = new DummyElasticSearchQuery();
+        final DiseaseUmlsHypernymQueryDecorator decorator = new DiseaseUmlsHypernymQueryDecorator(dummyQuery);
+        final TopicSet topicSet = new TopicSet(new File("src/main/resources/topics/topics2018.xml"), Challenge.TREC_PM, Task.PUBMED, 2017);
+        for (Topic topic : topicSet.getTopics()) {
+            decorator.expandTopic(topic);
+            if (topic.getDisease().equals("melanoma"))
+                assertThat(topic.getDiseaseHypernyms()).contains("neoplasm of skin", "melanocytes", "tumor", "disease of epidermal appendage");
```
Comment (michelole) [inline src/test/java/at/medunigraz/imi/bst/trec/query/DiseaseUmlsHypernymQueryDecoratorTest.java]: Lexigram provides a totally different set here, but I actually prefer UMLS, as it has "neoplasm of skin", probably the most important relationship.

However, "tumor" should be filtered out somewhere (it is when using the `WordRemovalQueryDecorator`).

Comment (michelole) [reply src/test/java/at/medunigraz/imi/bst/trec/query/DiseaseUmlsHypernymQueryDecoratorTest.java]: Any clue how many hypernyms were found here? 10s? 100s? 1000s?

File: src/main/java/de/julielab/ir/umls/UmlsRelationsProvider.java (hunk 2 above)
Comment (khituras) [reply src/main/java/de/julielab/ir/umls/UmlsRelationsProvider.java]: It actually can be. It was only public for the unit tests.
However, the workaround now is that there are static fields which can be set by the tests. But that means they could also be set in production code which should happen. So this is a fragile thing.
The issue is: How to properly configure a singletonf or testing?

File: src/test/java/at/medunigraz/imi/bst/trec/query/DiseaseUmlsHypernymQueryDecoratorTest.java (hunk 4 above)
Comment (khituras) [reply src/test/java/at/medunigraz/imi/bst/trec/query/DiseaseUmlsHypernymQueryDecoratorTest.java]: Yes, same as "disorder".

**Comment by michelole** [comment]
`src/test/resources/umls/melanoma.synset` with 2k lines is in the commit, was that intended?

**Comment by khituras** [comment]
> `src/test/resources/umls/melanoma.synset` with 2k lines is in the commit, was that intended?

Yes, a lot of synsets for melanoma ;-) I didn't really do much filtering, I think its OK this way.

**Comment by michelole** [comment]
Next week I can run some experiments comparing the impact on IR metrics of: Lexigram _versus_ UMLS _versus_  word embeddings.

**Comment by michelole** [comment]
> > `src/test/resources/umls/melanoma.synset` with 2k lines is in the commit, was that intended?
> 
> Yes, a lot of synsets for melanoma ;-) I didn't really do much filtering, I think its OK this way.

Is that subset of UMLS public?

**Comment by khituras** [comment]
Whew OK you got me at a lot of points. The issue is, I am not ready to invest more time into this test. I will do something purely synthetical. Obviously, I prefer more real data but I see the problems.

File: src/main/java/de/julielab/ir/umls/UmlsRelationsProvider.java (hunk 2 above)
Comment (michelole) [reply src/main/java/de/julielab/ir/umls/UmlsRelationsProvider.java]: You can use package-private (don't define the accessor level) and have the tests in the same package. Or, you can extend the class in the test and define a public constructor.

**Pull request merged**